index.js

```diff
@@ -413,8 +417,15 @@ function union(types, name) {
 
   Union.dispatch = function (x) { // default dispatch implementation
     for (var i = 0, len = types.length; i < len; i++ ) {
-      if (is(x, types[i])) {
-        return types[i];
+      var type = types[i];
+      if (isUnion(type)) {
+        var t = type.dispatch(x);
+        if (!isNil(t)) {
```

Comment:
It is, because if the concrete type isn't find it this particular union, it won't check the next union type in the array.